### PR TITLE
Fix missing IT equip and other gains in zone component load summary, room air models, and zone gain outputs

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
Addresses #6146 and more.
1. Gains from ElectricEquipment:ITE:AirCooled now show in the Zone Component Loads Summary report.  
2. They are also included in the occupied zone gains for room air models.  
3. And also included now in output variables "Zone Total Internal Convective Heating Energy" and " . . . Rate".

- [x] Need a unit test yet for this.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by rraustad** [comment]
Why wouldn't something like this show diff's somewhere? Not even a unit test? Are these just used for warnings?

**Comment by rraustad** [comment]
Before and after result comparing Develop to this branch. Is this the expected answer? No Sensible-Delayed? Blank Sensible-Return?

**Comment by rraustad** [comment]
Lights and refrigeration are the only components that will report a sensible load to return air. @EnergyArchmage should there be a delayed sensible component here?

**Comment by rraustad** [comment]
It appears that a delay occurs only for equipment with a radiant component to surfaces. So Sensible-Delayed should be 0. I looked at the ITE object and saw the fraction to zone losses input and changed that from 1 to 0.9. I was expecting to see Sensible-Instant drop by 10% but it did not. It did drop but only by 1%.

  ElectricEquipment:ITE:AirCooled,
    WestDataCenter_Equip,    !- Name
    West Zone,               !- Zone Name
    0.9,                       !- Fraction of Electric Power Supply Losses to Zone

6146_point9losses

**Comment by rraustad** [comment]
Here is where that fraction loss to zone (UPSLossToZoneFrac) is used:

    UPSPower = ( CPUPower + FanPower ) * max(( 1.0 - ZoneITEq( Loop ).DesignUPSEfficiency * CurveValue( ZoneITEq( Loop ).UPSEfficFPLRCurve, UPSPartLoadRatio ) ) , 0.0 );
    UPSHeatGain = UPSPower * ZoneITEq( Loop ).UPSLossToZoneFrac;

And then this calculation puts this heat to the zone and includes CPUPower and FanPower again? This looks like it adds ( 1 + UPSLossToZoneFrac ) heat gain to the zone.

    ElectricEquipment:ITE:AirCooled,
      WestDataCenter_Equip,    !- Name
      West Zone,               !- Zone Name
      AdjustedSupply,          !- Air Inlet Connection Type (i.e., AirConnection == ITEInletAdjustedSupply)

    if ( AirConnection == ITEInletAdjustedSupply || AirConnection == ITEInletZoneAirNode ) {
        // If not a room air model, then all ITEquip power input is a convective heat gain to the zone heat balance, plus UPS heat gain
        ZoneITEq( Loop ).ConGainRateToZone = CPUPower + FanPower + UPSHeatGain;
    } else if ( AirConnection == ITEInletRoomAirModel ) {
        // Room air model option not implemented yet - set room air model outlet node conditions here
        // If a room air model, then the only convective heat gain to the zone heat balance is the UPS heat gain
        ZoneITEq( Loop ).ConGainRateToZone = UPSHeatGain;
    }

**Comment by rraustad** [comment]
OK, this is correct. The UPS efficiency is 0.9. So 10% of the CPU + Fan power is a loss from the UPS power supply to the zone. And when 0.9 is used for Fraction of Electric Power Supply Losses to Zone, then the change in Sensible - Instant should be 10% * 10% = 1% as shown above.

    ElectricEquipment:ITE:AirCooled,
      WestDataCenter_Equip,    !- Name
      West Zone,               !- Zone Name
      0.9,                     !- Design Electric Power Supply Efficiency

**Comment by rraustad** [comment]
Document change in Zone Total Internal Convective Heating Rate/Energy:

itebranch

**Comment by rraustad** [comment]
@EnergyArchmage if 90% of losses go to the zone, where does the other 10% go? Assumed lost to outdoors?

ElectricEquipment:ITE:AirCooled,
0.9, !- Fraction of Electric Power Supply Losses to Zone

**Comment by mjwitte** [comment]
@rraustad @EnergyArchmage @JasonGlazer It is a little concerning that a new end-use can be left out of these reports without any warning to the developer.  Is there some way we could check for unused internal gains across these reports?

**Comment by JasonGlazer** [comment]
I'm a little surprised. We have a warning message that gets generated in WriteBEPSTable if the end uses don't add up to the total:  "In the Annual Building Utility Performance Summary Report the total row does not match the sum of the column for: " + resource. see line  

https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/OutputReportTabular.cc#L7721

**Comment by mjwitte** [comment]
That's a different set of sub-parts.  This is [internal gain types](https://github.com/NREL/EnergyPlus/blob/3ef1eaf299a4db7fb70d10ade1c7b993d5c694ff/src/EnergyPlus/DataHeatBalance.cc#L256-L308).

File: tst/EnergyPlus/unit/InternalHeatGains.unit.cc [hunk 1]
```diff
@@ -441,3 +446,70 @@ TEST_F(EnergyPlusFixture, InternalHeatGains_ElectricEquipITE_BeginEnvironmentRes
...
+
+	// Legitimate gain types excluded from this total
+	expectedTotConvGains -= convGains( DataHeatBalance::IntGainTypeOf_ZoneContaminantSourceAndSinkCarbonDioxide ); // this is only used for CO2
+	expectedTotConvGains -= convGains( DataHeatBalance::IntGainTypeOf_ZoneContaminantSourceAndSinkGenericContam ); // this is only used for generic contaminants
+	expectedTotConvGains -= convGains( DataHeatBalance::IntGainTypeOf_DaylightingDeviceTubular ); // this should be included in the component loads as Fenestration Solar or Fenestration Conduction - Instant?
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/InternalHeatGains.unit.cc]: @JasonGlazer It doesn't appear that conduction from tubular daylighting devices is included in the zone component loads summary.  It's not in the convective gains part, but maybe it's already included somewhere else?

File: src/EnergyPlus/InternalHeatGains.cc [hunk 2]
```diff
@@ -5341,10 +5341,10 @@ namespace InternalHeatGains {
...
+		static Array1D_int IntGainTypesHvacLoss( 20, { IntGainTypeOf_ZoneBaseboardOutdoorTemperatureControlled, IntGainTypeOf_ThermalStorageChilledWaterMixed, IntGainTypeOf_ThermalStorageChilledWaterStratified, IntGainTypeOf_PipeIndoor, IntGainTypeOf_Pump_VarSpeed, IntGainTypeOf_Pump_ConSpeed, IntGainTypeOf_Pump_Cond, IntGainTypeOf_PumpBank_VarSpeed, IntGainTypeOf_PumpBank_ConSpeed, IntGainTypeOf_PlantComponentUserDefined, IntGainTypeOf_CoilUserDefined, IntGainTypeOf_ZoneHVACForcedAirUserDefined, IntGainTypeOf_AirTerminalUserDefined, IntGainTypeOf_PackagedTESCoilTank, IntGainTypeOf_FanSystemModel, IntGainTypeOf_SecCoolingDXCoilSingleSpeed, IntGainTypeOf_SecHeatingDXCoilSingleSpeed, IntGainTypeOf_SecCoolingDXCoilTwoSpeed, IntGainTypeOf_SecCoolingDXCoilMultiSpeed, IntGainTypeOf_SecHeatingDXCoilMultiSpeed } );
```
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc:5349]: @JasonGlazer Also note several additions here for Equip, Refrig, and HvacLoss.  Please confirm if these look correct.

Comment (JasonGlazer) [reply src/EnergyPlus/InternalHeatGains.cc:5349]: Those additions seem reasonable.

File: tst/EnergyPlus/unit/InternalHeatGains.unit.cc (hunk 1 above)
Comment (JasonGlazer) [reply tst/EnergyPlus/unit/InternalHeatGains.unit.cc]: I thought that TDD were treated more like a surface than an internal gain.

Comment (mjwitte) [reply tst/EnergyPlus/unit/InternalHeatGains.unit.cc]: The diffuser is treated as a surface, but apparently the gain from the tube to transition zone(s) is added as convective heat gain component.
```
SetupZoneInternalGain( TDDPipe( PipeNum ).TZone( TZoneNum ), "DaylightingDevice:Tubular", TDDPipe( PipeNum ).Name, IntGainTypeOf_DaylightingDeviceTubular, TDDPipe( PipeNum ).TZoneHeatGain( TZoneNum ) )`

// Distribute absorbed solar gain in proportion to transition zone length
TDDPipe( PipeNum ).TZoneHeatGain( TZoneNum ) = TotTDDPipeGain * ( TDDPipe( PipeNum ).TZoneLength( TZoneNum ) / TDDPipe( PipeNum ).TotLength );

  DaylightingDevice:Tubular,
    Pipe1,                   !- Name
    Dome1,                   !- Dome Name
    Diffuser1,               !- Diffuser Name
    TDD Pipe,                !- Construction Name
    0.3556,                  !- Diameter {m}
    1.4,                     !- Total Length {m}
    0.28,                    !- Effective Thermal Resistance {m2-K/W}
    Daylit Attic Zone,       !- Transition Zone 1 Name
    1.1;                     !- Transition Zone 1 Length {m}
```

**Comment by mjwitte** [comment]
Added the tubular daylighting device conduction to "Fenestration Conduction" in the zone component load summary tables.  Here's results from a modified version of DaylightingDeviceTubular (added HVAC and autosizing, this file in in DefectFiles now).
Before
image
image
image

After
image
image

**Comment by rraustad** [comment]
Where did these numbers come from? 56.53 W, -5.7 W ? How do we know if all loads are included in reports? You didn't change the original fix, you added new code to it. How do we know what else is needed (besides your brilliance)?

    -		static Array1D_int TradIntGainTypes( 7, { IntGainTypeOf_People, IntGainTypeOf_Lights, IntGainTypeOf_ElectricEquipment, IntGainTypeOf_GasEquipment, IntGainTypeOf_HotWaterEquipment, IntGainTypeOf_SteamEquipment, IntGainTypeOf_OtherEquipment } );
     +		static Array1D_int TradIntGainTypes( 8, { IntGainTypeOf_People, IntGainTypeOf_Lights, IntGainTypeOf_ElectricEquipment, IntGainTypeOf_ElectricEquipmentITEAirCooled, IntGainTypeOf_GasEquipment, IntGainTypeOf_HotWaterEquipment, IntGainTypeOf_SteamEquipment, IntGainTypeOf_OtherEquipment } );

**Comment by JasonGlazer** [comment]
Those numbers are the differences between the sum of the estimated component loads and the load calculated by EnergyPlus. The estimation procedure is outlined in the Engineering Reference chapter 24 and is approximate since the heat balance algorithms in EnergyPlus do not normally determine the component loads. They are an indicator of how well the estimate works and do not necessarily indicate if a component is missing. That would have to be done some other way.

**Comment by mjwitte** [comment]
At this point, the new unit test is the only check on whether the component load summary is including all possible IntGainTypeOf* loads.  Beyond that, for other places that use these types of gains to grab subtotals, there is no check.

The good news is that the main heat balance accumulates *all* convective IntGains, and *all* radiant IntGains, etc. without any filters, so that part is OK.  `SumAllInternalConvectionGains` in ZoneTempPredictorCorrector::CalcZoneSums.  `SumAllInternalRadiationGains` and `SumAllInternalLatentGains` in InternalHeatGains::InitInternalHeatGains.

**Comment by Myoldmopar** [comment]
@rraustad Thanks for digging in on this and starting the conversation above that revealed issues with the underlying behavior.  It's a little hard to tell from the conversation whether there are still open questions about the implementation.  I don't mind diving in to review, but if there are still open things I won't do this one yet.

**Comment by rraustad** [comment]
@Myoldmopar  This looks ready to go. The only remaining item is beyond the scope of this change and regards @mjwitte comment so that we don't miss including future internal gains in the end-use reports:

` It is a little concerning that a new end-use can be left out of these reports without any warning to the developer. Is there some way we could check for unused internal gains across these reports?`

File: tst/EnergyPlus/unit/InternalHeatGains.unit.cc [hunk 3]
```diff
@@ -441,3 +446,70 @@ TEST_F(EnergyPlusFixture, InternalHeatGains_ElectricEquipITE_BeginEnvironmentRes
...
+
+	// Legitimate gain types excluded from this total
+	expectedTotConvGains -= convGains( DataHeatBalance::IntGainTypeOf_ZoneContaminantSourceAndSinkCarbonDioxide ); // this is only used for CO2
+	expectedTotConvGains -= convGains( DataHeatBalance::IntGainTypeOf_ZoneContaminantSourceAndSinkGenericContam ); // this is only used for generic contaminants
+	expectedTotConvGains -= convGains( DataHeatBalance::IntGainTypeOf_DaylightingDeviceTubular ); // this is included in Fenestration Conduction - Sensible Instant
+
+	// ** NOTE: If this unit test fails, the likely cause is that a new internal gain type was added, but it was not added to one of the subtotal types in InternalHeatGains::GatherComponentLoadsIntGain()
+	// this also means that the new type may be missing from other places that collect internal gains by subgroups, such as the room air models and output reporting for zone-level gains
+	// search for IntGainTypeOf_Lights for places where these types of subtotals occur and add the new type as appropriate
+	EXPECT_EQ( totConvGains, expectedTotConvGains );
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/InternalHeatGains.unit.cc:509]: This is a great unit test.

File: src/EnergyPlus/UFADManager.cc [hunk 4]
```diff
@@ -1053,7 +1053,7 @@ namespace UFADManager {
...
...t IntGainTypesOccupied( 29, { IntGainTypeOf_People, IntGainTypeOf_WaterHeaterMixed, IntGainTypeOf_WaterHeaterStratified, IntGainTypeOf_ThermalStorageChilledWaterMixed, IntGainTypeOf_ThermalStorageChilledWaterStratified, IntGainTypeOf_ElectricEquipment, IntGainTypeOf_ElectricEquipmentITEAirCooled, IntGainTypeOf_GasEquipment, IntGainTypeOf_HotWaterEquipment, IntGainTypeOf_SteamEquipment, IntGainTypeOf_OtherEquipment, IntGainTypeOf_ZoneBaseboardOutdoorTemperatureControlled, IntGainTypeOf_GeneratorFuelCell, IntGainTypeOf_WaterUseEquipment, IntGainTypeOf_GeneratorMicroCHP, IntGainTypeOf_ElectricLoadCenterTransformer, IntGainTypeOf_ElectricLoadCenterInverterSimple, IntGainTypeOf_ElectricLoadCenterInverterFunctionOfPower, IntGainTypeOf_ElectricLoadCenterInverterLookUpTable, IntGainTypeOf_ElectricLoadCenterStorageBattery, IntGainTypeOf_ElectricLoadCenterStorageSimple, IntGainTypeOf_PipeIndoor, IntGainTypeOf_RefrigerationCase, IntGainTypeOf_RefrigerationCompressorRack, IntGainTypeOf_RefrigerationSystemAirCooledCondenser, IntGainTypeOf_RefrigerationSystemSuctionPipe, IntGainTypeOf_RefrigerationSecondaryReceiver, IntGainTypeOf_RefrigerationSecondaryPipe, IntGainTypeOf_RefrigerationWalkIn } );
```
Comment (Myoldmopar) [inline src/EnergyPlus/UFADManager.cc:1056]: Sometime these should be made into multiple lines I think, would make it easier to diff.

File: src/EnergyPlus/OutputReportTabular.cc [hunk 5]
```diff
@@ -11448,6 +11450,11 @@ namespace OutputReportTabular {
 				// convect that heat that timestep.
 				//feneSolarInstantSeq(ZoneNum,TimeStepInDay,CurOverallSimDay) = 0
 			}
+			for ( int izone = 1; izone <= NumOfZones; ++ izone ){
+				Real64 tubularGain = 0.0;
+				InternalHeatGains::SumInternalConvectionGainsByTypes( izone, IntGainTypesTubular, tubularGain );
+				feneCondInstantSeq( CurOverallSimDay, TimeStepInDay, izone ) += tubularGain;
+			}
```
Comment (Myoldmopar) [inline src/EnergyPlus/OutputReportTabular.cc:11457]: :+1:

**Review by Myoldmopar — APPROVED** [review]

File: src/EnergyPlus/UFADManager.cc (hunk 4 above)
Comment (mjwitte) [reply src/EnergyPlus/UFADManager.cc:1056]: I can do that (multiple lines thing) here if you like.

Comment (Myoldmopar) [reply src/EnergyPlus/UFADManager.cc:1056]: Unless there are actual meaningful code changes, I'd prefer to not bother with waiting on CI to make another pass.

**Comment by Myoldmopar** [comment]
Built this branch and compared the results of the 2Zone example file model, with     ZoneComponentLoadSummary turned on in the IDF.  Confirmed the updated values showing up in the `Sensible - Instant [W]` and thus the `Total` columns of the component load summary table.  This is good to go.  I'll merge.

**Pull request merged**